Consider a dataset contains 90 responses for 14 different variables that customers consider while purchasing car. 

# Step1:.  Evaluate factorability!

In [1]:
#import libraries
import pandas as pd
import numpy as np
from sklearn.decomposition import FactorAnalysis
from factor_analyzer import FactorAnalyzer
import matplotlib.pyplot as plt


In [2]:
#step1(b)
data=pd.read_csv("D:/Data visualization/Dr Mwendia_Class/carpurchase_factors.csv")
#a way of listing attributes
attributes=data.columns

In [3]:
print(data)

    Price  Safety  Exterior_Looks  Space_comfort  Technology  \
0       4       4               5              4           3   
1       3       5               3              3           4   
2       4       4               3              4           5   
3       4       4               4              3           3   
4       5       5               4              4           5   
..    ...     ...             ...            ...         ...   
85      4       4               3              4           5   
86      4       3               4              4           5   
87      4       3               4              4           4   
88      4       4               3              4           4   
89      4       5               3              4           3   

    After_Sales_Service  Resale_Value  Fuel_Type  Fuel_Efficiency  Color  \
0                     4             5          4                4      2   
1                     4             3          4                3      4   
2  

In [4]:
#Determine Factorability
#use Barlett_sphericity Test
from factor_analyzer.factor_analyzer import calculate_bartlett_sphericity
chi_square_value,p_value=calculate_bartlett_sphericity(data)
chi_square_value,p_value

(260.771089474057, 8.93220098429307e-19)

In this example, Bartlett ’s test indicates that the p-value is less than 0.05. Therefore the test was statistically significant, indicating that the observed correlation matrix is not an identity matrix and Factor analysis can be conducted.


In [5]:
#Determine Factorability
#perform KMD test
from factor_analyzer.factor_analyzer import calculate_kmo
kmo_all,kmo_model=calculate_kmo(data)
print('\nKMO Model',kmo_model)


KMO Model 0.6091437123668768


The overall KMO for our data is 0.609, which is mediocre.
This value indicates that we can proceed with your planned factor analysis.


# Step 2: Extract  Factors!

In [6]:
#Extract the factors
factor=FactorAnalysis().fit(data)
Factors=pd.DataFrame(factor.components_,columns=attributes)
print(Factors)

       Price    Safety  Exterior_Looks  Space_comfort  Technology  \
0  -0.261282  0.105198        0.081019      -0.093802   -0.140787   
1   0.063251 -0.168130       -0.184186      -0.379449   -0.268080   
2   0.124918  0.124247        0.066854       0.137942    0.117193   
3   0.073006 -0.140767        0.344325       0.011100   -0.075386   
4  -0.080811  0.064612        0.004853      -0.054486   -0.100244   
5   0.065501  0.042757       -0.045711       0.015591   -0.177875   
6   0.000000  0.000000       -0.000000       0.000000   -0.000000   
7   0.000000 -0.000000       -0.000000       0.000000   -0.000000   
8   0.000000  0.000000        0.000000      -0.000000    0.000000   
9   0.000000  0.000000        0.000000      -0.000000    0.000000   
10 -0.000000  0.000000        0.000000       0.000000   -0.000000   
11 -0.000000  0.000000        0.000000      -0.000000    0.000000   
12  0.000000 -0.000000       -0.000000      -0.000000   -0.000000   
13  0.000000  0.000000        0.00

In [12]:
#choose Factors
#Using Kaiser criterion_Get eigen values

fa=FactorAnalyzer(rotation=None)
fa.fit(data)
FactorAnalyzer(bounds=(0.005, 1), impute='median', is_corr_matrix=False,
               method='minres', n_factors=3, rotation=None, rotation_kwargs={},
               use_smc=True)

#check Eigenvalues



FactorAnalyzer(bounds=(0.005, 1), impute='median', is_corr_matrix=False,
               method='minres', n_factors=3, rotation=None, rotation_kwargs={},
               use_smc=True)

In [13]:
fa.loadings_

array([[ 0.4173656 , -0.17668348, -0.20536068],
       [-0.07901537,  0.31749456, -0.176638  ],
       [-0.04664407,  0.24116654,  0.00355491],
       [ 0.38115015,  0.72455861, -0.18249336],
       [ 0.24609196,  0.25505457, -0.07324372],
       [ 0.44436381,  0.27146377, -0.12708067],
       [ 0.40220556, -0.51972323, -0.22100494],
       [ 0.28310253,  0.41305672, -0.26028371],
       [ 0.68191509, -0.11410996,  0.09984058],
       [ 0.37676588, -0.25476735,  0.25798852],
       [ 0.59050535, -0.31484465, -0.06794407],
       [ 0.30787321,  0.16883587,  0.25244077],
       [ 0.49560538,  0.03252678,  0.2045716 ],
       [ 0.11179477,  0.30620795,  0.66441765]])